services/config/src/useDate.ts

```diff
@@ -0,0 +1,134 @@
+        Object.setPrototypeOf(this, DHIS2Date.prototype)
+    }
+
+    private _getTimeComponents(date: Date): DateComponents {
+        const year = date.getFullYear().toString().padStart(4, '0')
+        const month = (date.getMonth() + 1).toString().padStart(2, '0')
+        const days = date.getDate().toString().padStart(2, '0')
+        const hours = date.getHours().toString().padStart(2, '0')
+        const minutes = date.getMinutes().toString().padStart(2, '0')
+        const seconds = date.getSeconds().toString().padStart(2, '0')
+        const milliseconds = date.getMilliseconds().toString().padStart(3, '0')
+        return { year, month, days, hours, minutes, seconds, milliseconds }
+    }
+
+    private _getISOString(date: Date): string {
+        const { year, month, days, hours, minutes, seconds, milliseconds } =
+            this._getTimeComponents(date)
+        return `${year}-${month}-${days}T${hours}:${minutes}:${seconds}.${milliseconds}`
+    }
+
+    public getServerISOString(): string {
+        const serverDate = new Date(this.getTime() - this.serverOffset)
+        return this._getISOString(serverDate)
+    }
+
+    public getClientISOString(): string {
+        return this._getISOString(this)
+    }
+}
+
+const useServerTimeOffset = (serverTimezone: string): number => {
+    return useMemo(() => {
+        try {
+            const nowClientTime = new Date()
+            nowClientTime.setMilliseconds(0)
+
+            // 'sv' is used for localeString because it is the closest to ISO format
+            // in principle, any locale should be parsable back to a date, but we encountered an error
+            // when using en-US in certain environments, which we could not replicate when using 'sv'
+            // Converting to localeString and then back to date is unfortunately the only current way
+            // to construct a date that accounts for timezone.
+            const serverLocaleString = nowClientTime.toLocaleString('sv', {
+                timeZone: serverTimezone,
+            })
+            const nowServerTimeZone = new Date(serverLocaleString)
+            nowServerTimeZone.setMilliseconds(0)
+
+            return nowClientTime.getTime() - nowServerTimeZone.getTime()
+        } catch (err) {
+            console.error(
+                'Server time offset could not be determined; assuming no client/server difference',
+                err
+            )
+            // if date is not constructable with timezone, assume 0 difference between client/server
+            return 0
+        }
+    }, [serverTimezone])
+}
+
+export const useDate = (): {
+    fromServerDate: (date?: DateInput) => DHIS2Date
+    fromClientDate: (date?: DateInput) => DHIS2Date
+} => {
+    const { systemInfo } = useContext(ConfigContext)
+    const clientTimezone = Intl.DateTimeFormat().resolvedOptions().timeZone
+    const { serverTimeZoneId: serverTimezone } = systemInfo || {
+        serverTimeZoneId: clientTimezone,
+    }
```

Comment:
I agree that in principle the `clientTimeZone` is a sensible fallback, since we then assume both client and server are in the same time zone, which is probably the soundest assumption one could make. And probably causing the whole app to crash due to something small like this is also not what we want....

On the other hand, the `systemInfo` object _should always_ include a `serverTimeZoneId`, and if this is not available something is wrong and timezone based date conversion could be going wrong in the app. We can improve DX by showing _something_ on the console. Either a `console.error()` or `console.warn()` would do.